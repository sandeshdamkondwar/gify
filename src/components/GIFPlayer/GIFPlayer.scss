$play-button-height: 45px;
$gif_btn_bg_base_color: #000 !default;
$gif_btn_bg_opacity: 0.5 !default;
$gif_btn_bg_opacity_hover: 0.7 !default;
$gif_btn_bg_color: rgba($gif_btn_bg_base_color, $gif_btn_bg_opacity) !default;
$gif_btn_bg_color_hover: rgba(
  $gif_btn_bg_base_color,
  $gif_btn_bg_opacity_hover
) !default;

$gif_btn_box_shadow_base_color: #000 !default;
$gif_btn_box_shadow_opacity: $gif_btn_bg_opacity !default;
$gif_btn_box_shadow_opacity_hover: $gif_btn_bg_opacity_hover !default;
$gif_btn_box_shadow_color: rgba(
  $gif_btn_box_shadow_base_color,
  $gif_btn_box_shadow_opacity
) !default;
$gif_btn_box_shadow_color_hover: rgba(
  $gif_btn_box_shadow_base_color,
  $gif_btn_box_shadow_opacity_hover
) !default;

$gif_btn_text_color: #fff !default;

$gif_btn_border_width: 2px !default;
$gif_btn_border_style: dashed !default;
$gif_btn_border_color: $gif_btn_text_color !default;

$gif_btn_regular_opacity: 1 !default;
$gif_btn_playing_opacity: 0.5 !default;

$gif_btn_paused_rotation: 0deg !default;
$gif_btn_playing_rotation: 180deg !default;

$gif_btn_transition_duration: 0.4s;

.gif_player {
  display: inline-block;
  position: relative;
  user-select: none;
  -webkit-touch-callout: none;
  -webkit-tap-highlight-color: transparent;

  .play_button {
    width: $play-button-height;
    height: $play-button-height;
    background-color: $gif_btn_bg_color;
    border: $gif_btn_border_width $gif_btn_border_style $gif_btn_border_color;
    border-radius: 50%;
    box-shadow: 0 0 0 3px $gif_btn_box_shadow_color;
    color: $gif_btn_text_color;
    cursor: pointer;

    left: 50%;
    opacity: $gif_btn_regular_opacity;
    position: absolute;
    top: 50%;
    transform: translate(-50%, -50%) scale(1) rotate($gif_btn_paused_rotation);
    transition: transform $gif_btn_transition_duration,
      opacity $gif_btn_transition_duration;

    &:hover {
      background-color: $gif_btn_bg_color_hover;
      box-shadow: 0 0 0 3px $gif_btn_box_shadow_color_hover;

      &::after {
        font-size: 0.9em;
      }
    }

    &::after {
      font-size: 1em;
      content: "GIF";
      width: 45px;
      display: inline-block;
      text-align: center;
      vertical-align: middle;
      line-height: $play-button-height;
      font-weight: bold;
    }
  }

  &.playing .play_button {
    transform: translate(-50%, -50%) scale(0) rotate($gif_btn_playing_rotation);
    opacity: $gif_btn_playing_opacity;
  }

  img {
    width: 100%;
  }
}
