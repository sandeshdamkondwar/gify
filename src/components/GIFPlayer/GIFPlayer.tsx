import React, { useState, useCallback } from "react";
import classNames from "classnames";

const PADDING = 5;
interface IGIFPlayerProps {
  gif: string;
  still: string;
  height: number;
  title: string;
  className?: string;
  single?: boolean;
  autoPlay?: boolean;
}

const GifPlayer = ({
  gif,
  still,
  title,
  className,
  single,
  height,
  autoPlay = false,
}: IGIFPlayerProps) => {
  const refPlaceholder = React.useRef<HTMLDivElement>(null);
  const [playing, setplaying] = useState(autoPlay);

  const toggle = useCallback(() => {
    if (autoPlay) return;
    setplaying(!playing);
  }, [autoPlay, playing]);

  const removePlaceholder = useCallback(() => {
    refPlaceholder?.current?.remove();
  }, []);

  const classes = classNames("gif_player", className || "", {
    playing: playing,
    single: single,
  });

  let imageAttr: any = {
    src: playing ? gif || still : still || gif,
    onLoad: removePlaceholder,
    onError: removePlaceholder,
  };

  if (single) {
    imageAttr.key = imageAttr.src;
  }

  const imageHeight = Math.floor(height - PADDING * 2);
  imageAttr.height = imageHeight;

  return (
    <div
      className={classes}
      onClick={toggle}
      style={{ height: `${imageHeight}px` }}
    >
      <div className="play_button" data-test-id={title}></div>

      <div className="placeholder" ref={refPlaceholder}></div>
      <img data-testid="gif-image" {...imageAttr} alt={title} />
    </div>
  );
};

export default GifPlayer;
