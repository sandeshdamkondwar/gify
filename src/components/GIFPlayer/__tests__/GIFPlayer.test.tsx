import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import GIFPlayer from "../GIFPlayer";

const TestComponent = (
  <GIFPlayer
    gif="https://media0.giphy.com/media/13dPffliPA4qbTTb1m/200w.gif?cid=9a7f325649d1c3adcb43a797139887d707ab1f9bc10642d2&rid=200w.gif"
    still="https://media0.giphy.com/media/13dPffliPA4qbTTb1m/200w.gif?cid=9a7f325649d1c3adcb43a797139887d707ab1f9bc10642d2&rid=200w.gif"
    height={266}
    title="test image"
  />
);

// Snapshot test cases
it("matches the snapshot", () => {
  const { baseElement } = render(TestComponent);
  expect(baseElement).toMatchSnapshot();
});
