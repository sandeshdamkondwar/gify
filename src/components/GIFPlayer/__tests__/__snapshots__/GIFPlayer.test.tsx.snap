// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<body>
  <div>
    <div
      class="gif_player"
      style="height: 256px;"
    >
      <div
        class="play_button"
        data-test-id="test image"
      />
      <div
        class="placeholder"
      />
      <img
        alt="test image"
        data-testid="gif-image"
        height="256"
        src="https://media0.giphy.com/media/13dPffliPA4qbTTb1m/200w.gif?cid=9a7f325649d1c3adcb43a797139887d707ab1f9bc10642d2&rid=200w.gif"
      />
    </div>
  </div>
</body>
`;
